### ISSUE TITLE
delete!(::Dict{A, B}, x) doesn't throw an error when x is not of type A

### ISSUE THREAD

**Issue by colinfang** [JuliaLang/julia#15614]
```julia
x = Dict{Int, Char}(12=>'c')
delete!(x, "kk")
```

I am expecting an error.

**Comment by nalimilan** [comment]
This behavior is consistent with the fact that the following doesn't raise an error either:
```julia
julia> x = Dict{Int, Char}(12=>'c')
Dict(12=>'c')

julia> "kk" in keys(x)
false
```

**Comment by StefanKarpinski** [comment]
I've come to favor a more strict behavior with regard to typed collections and keys, where, before checking for the existence of a value in a collection, we convert it to the expected type and check that it remains `isequal` to the original value. If that's not the case, then you're checking for an "unreasonable value" for that collection and it would be best to throw an error. There are cases where one is using a collection that it is reasonable to look for the existence of any kind of value in, but in that case, I think that it makes sense to use a collection with an element type of `Any`.

**Comment by nalimilan** [comment]
Makes sense, but it means that `in` would no longer follow `==`/`isequal` semantics. Or should we also change the behavior of the latter?

**Comment by JeffBezanson** [comment]
Arguably `delete!` should do the same convert/isequal check that `setindex!` already does for Dicts.

I'm less sure about extending this to lookup. Right now, Dict has the following property: if a Dict only ever has keys of type K, and you only try to insert keys of type K, then its behavior is exactly the same no matter what the container type is (as long as its type is some supertype of K of course). This makes it safe to narrow the type of a Dict. If the container type determined which lookups are allowed, we would lose this property.

For example, say we are dealing with `Rational{Int}`s. If I notice one of my Dicts happens to only ever have `Int`-valued keys, it's safe to make the key type `Int`, even though I might be looking for non-integer-valued Rationals in it sometimes.

**Comment by cossio** [comment]
I am in favor of throwing *different* kind of errors. If the type of `k` in `d[k]` is not the same as the type of keys in the dictionary, then the error should say that. A "key not found" is too generic.

https://discourse.julialang.org/t/dict-key-not-found-instead-of-type-error/4972
